### DISCUSSION TITLE
slice image based on line

### DISCUSSION THREAD

**Discussion by hmd-shrf** [ImageMagick/ImageMagick#2104] [Help] [answered]
Hi, (my first post here)

Hope someone can help me on this or just to confirm if this is actually doable or not.

I've a PNG file that has horizontal lines. I'd like to split the image into smaller files based on these lines.
Is this possible using imageMagick?

I'm attaching the files as well so that it is more clear.

Thanks

Original file:
![Original](https://user-images.githubusercontent.com/5032598/83739548-2c1e8600-a66f-11ea-8300-718e4f82fa0d.png)

Section 1:
<img width="827" alt="1" src="https://user-images.githubusercontent.com/5032598/83739587-3c366580-a66f-11ea-809d-237b0645b894.png">

Section 2:
<img width="826" alt="2" src="https://user-images.githubusercontent.com/5032598/83739600-3fc9ec80-a66f-11ea-91a5-398416e45b5d.png">

Section 3:
<img width="827" alt="3" src="https://user-images.githubusercontent.com/5032598/83739612-435d7380-a66f-11ea-834e-bbdebaf31b66.png">

Section 4:
<img width="827" alt="4" src="https://user-images.githubusercontent.com/5032598/83739621-46586400-a66f-11ea-9232-d75067d6d6c8.png">

**Comment by snibgo** [comment]
It is possible, but it needs a shell script.

What version of IM, on what platform?

Do you want to remove the green lines, or what?

One approach, assuming IM v7 on Windows BAT, is:

```
magick ^
  83739548-2c1e8600-a66f-11ea-8300-718e4f82fa0d.png ^
  -channel Red -separate +channel ^
  -scale "1x^!" -threshold 20%% ^
  -define connected-components:verbose=true -connected-components 4 ^
  NULL:
```

The text output from that is:

```
Objects (id: bounding-box centroid area mean-color):
  6: 1x714+0+1625 0.0,1981.5 714 gray(255)
  2: 1x591+0+476 0.0,771.0 591 gray(255)
  4: 1x548+0+1072 0.0,1345.5 548 gray(255)
  0: 1x471+0+0 0.0,235.0 471 gray(255)
  5: 1x5+0+1620 0.0,1622.0 5 gray(0)
  3: 1x5+0+1067 0.0,1069.0 5 gray(0)
  1: 1x5+0+471 0.0,473.0 5 gray(0)
```

The four lines with "gray(255)" give the vertical spans you want. A shell script can read this text, extracting the H and Y parts of each bounding box. Then crop to those numbers.

**Reply by hmd-shrf** [reply]
Thank you Snibgo for the reply.

Umm, it's a Windows platform.
Version: ImageMagick 7.0.10-2 Q16 x64 2020-03-22

No, I don't need the lines in the sliced images.

Also, I tried the code and this is the result I get
```
Objects (id: bounding-box centroid area mean-color):
  0: 1x1+0+0 0.0,0.0 1 gray(0)
```
...not sure where I am going wrong.

I am going through the documentation and trying to understand each command used in the code above. By the way, the horizontal line is put there manually. Would the color used in the line or the size of the line make it easier to achieve this?

Thanks again.

**Reply by hmd-shrf** [reply]
> Thank you Snibgo for the reply.
> 
> Umm, it's a Windows platform.
> Version: ImageMagick 7.0.10-2 Q16 x64 2020-03-22
> 
> Also, I tried the code and this is the result I get
> 
> ```
> Objects (id: bounding-box centroid area mean-color):
>   0: 1x1+0+0 0.0,0.0 1 gray(0)
> ```
> 
> ...not sure where I am going wrong.
> 
> I am going through the documentation and trying to understand each command used in the code above. By the way, the horizontal line is put there manually. Would the color used in the line or the size of the line make it easier to achieve this?
> 
> Thanks again.

Got the text output as in the reply. Seems I had an error before. 

```
  6: 1x714+0+1625 0.0,1981.5 714 gray(255)
  2: 1x591+0+476 0.0,771.0 591 gray(255)
  4: 1x548+0+1072 0.0,1345.5 548 gray(255)
  0: 1x471+0+0 0.0,235.0 471 gray(255)
  5: 1x5+0+1620 0.0,1622.0 5 gray(0)
  3: 1x5+0+1067 0.0,1069.0 5 gray(0)
  1: 1x5+0+471 0.0,473.0 5 gray(0)
```

Which numbers from the output text would be the Y and height?

**Comment by snibgo** [comment, marked as answer]
"-channel Red -separate +channel" considers just the red channel.

"-scale" resizes the image to one pixel wide. 

In "1x714+0+1625", the numbers are WxH+X+Y, the width and height of the largest white rectangle, and the coordinate of the top-left corner. As the image is only one pixel wide, we ignore W and X. H is 714, so the white rectangle is 714 pixels high. It starts at y=1625 in the input image. So the crop will be "-crop x714+0+1625".

You can extract the numbers in a "for" loop, like this:

```
setlocal enabledelayedexpansion

set SRC=83739548-2c1e8600-a66f-11ea-8300-718e4f82fa0d.png

set Cnt=0

for /F "usebackq tokens=3,5 delims=x+: " %%A in (`%IMG7%magick ^
   %SRC% ^
  -channel Red -separate +channel ^
  -scale "1x^!" -threshold 20%% ^
  -define connected-components:verbose^=true -connected-components 4 ^
  NULL: ^| findstr gray^(255^)`) do (
  echo %%A %%B

  %IMG7%magick ^
    %SRC% ^
    -crop x%%A+0+%%B +repage ^
    x_!Cnt!.png

  set /A Cnt+=1
)
```
The results x_0.png, x_1.png etc are the parts of the input image that were separated by green lines.

**Reply by hmd-shrf** [reply]
Thank you!
When I ran the code (rechecked the code as well), I get only one file and it is named as "x_!Cnt!.png" which has the first section of the file. What could be the cause? 

Also, I think I understand the code now, all of the steps except for line that starts with -define.

**Reply by hmd-shrf** [reply]
I put the following code on top of script and it works now.

`setlocal EnableDelayedExpansion`

**Reply by hmd-shrf** [reply]
Is there a way to get the connected component data sorted on Y Ascending?

**Reply by snibgo** [reply]
IM can't sort by y-coordinate. You could do that in an external sort tool. More easily, the script I showed named the output files with a counter. Instead, you could use the y-coordinate, perhaps with leading zeros.

**Reply by hmd-shrf** [reply]
Excellent! Worked. Thank you @snibgo.

**Comment by snibgo** [comment]
Yes, sorry about that. My computer doesn't need that line so I often forget to put it in. I edited my post to insert that line.

On "-connected-components", see http://magick.imagemagick.org/script/connected-components.php

**Reply by hmd-shrf** [reply]
ImageMagick is awesome!
I hope this thread could be useful to someone else too.
Thank you Snibgo. Thank you for your help, time and contribution.
